src/lib/iconvenc/iconvenc.pas

```diff
@@ -0,0 +1,224 @@
+{
+    This file is part of the Free Pascal run time library.
+    Copyright (c) 2000 by Marco van de Voort(marco@freepascal.org)
+    member of the Free Pascal development team
+
+    libiconv header translation + a helper routine
+    http://wiki.freepascal.org/iconvenc
+
+    See the file COPYING.FPC, included in this distribution,
+    for details about the copyright. (LGPL)
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+}
+unit iconvenc;
+
+interface
+{$mode objfpc}{$H+}
+
+
+uses
+  ctypes,
+  {$IFNDEF MSWINDOWS}
+    unixtype,
+    baseunix,
+  {$ENDIF}
+  initc;
+
+{$ifdef beos}
+  const
+    ESysEILSEQ = EILSEQ;
+{$endif}
+
+
+{$IFDEF MSWINDOWS}
+  type
+    iconv_t = Pointer;
+
+  const
+    iconvlib = 'libiconv.dll';
+{$ELSE}
+  type
+    piconv_t = ^iconv_t;
+    iconv_t = pointer;
+
+    Ticonv_open = function(__tocode: pchar; __fromcode: pchar): iconv_t; cdecl;
+    Ticonv = function(__cd: iconv_t; __inbuf: ppchar; __inbytesleft: psize_t; __outbuf: ppchar; __outbytesleft: psize_t): size_t; cdecl;
+    Ticonv_close = function(__cd: iconv_t): cint; cdecl;
+{$ENDIF}
+
+{$if not defined(linux) and not defined(solaris)}  // Linux (and maybe glibc pl
+   {$if defined(haiku)}
+    {$linklib textencoding}
+   {$else}
+    {$if not defined(MSWINDOWS)}
+     {$linklib iconv}
+    {$endif}
+   {$endif}
+ {$define useiconv}
+{$endif linux}
+
+Const
+{$ifndef useiconv}
+  libiconvname='c';  // is in libc under Linux.
+{$else}
+  {$ifdef haiku}
+    libiconvname='textencoding';  // is in libtextencoding under Haiku
+  {$else}
+    libiconvname='iconv';
+  {$endif}
+{$endif}
+
+{$if (defined(darwin) and defined(cpupowerpc32)) or defined(haiku)}
+  iconvprefix='lib';
+{$else}
+  iconvprefix='';
+{$endif}
+
+{$IFDEF MSWINDOWS}
+  function iconv_open(ToCode, FromCode: PChar): iconv_t; cdecl; external iconvlib name 'libiconv_open';
+  function iconv(__cd: iconv_t; __inbuf: PPChar; var __inbytesleft: size_t; __outbuf: ppchar; var __outbytesleft: size_t): size_t; cdecl; external iconvlib name 'libiconv';
+  function iconv_close(cd: iconv_t): Integer; cdecl; external iconvlib name 'libiconv_close';
+
+  function errno_location: PInteger; cdecl; external 'msvcrt.dll' name '_errno';
```

Comment:
Further down non-Unix platforms define noerrnoiconv. Is errno_location used at all? I guess to make it work one would also have to define ESysEILSEQ and ESysE2BIG to the values used by the MXE toolchain.

Reply:
Doesn't seem like it's used, so I removed it along with the other unused code on Windows.

Reply:
It's was not used because they thought errno exists only on Unix. But our libiconv does use errno on Windows.

The code for Unix uses `cerrno` to get the value of errno. That's a property defined in the initc unit. Unfortunately [the implementation for win32](https://gitlab.com/freepascal.org/fpc/source/-/blob/main/rtl/win32/initc.pp) tries to use the function `__errno` from cygwin1.dll instead of `_errno` from msvcrt.dll. That won't work. I suggest we call errno_location directly instead of using cerrno. Just write `errno_location^` where `cerrno` was and remove `{$define noerrnoiconv}`.

Reply:
Hope I understood this correctly.

Changed the implementation in iconvenc back to the one used by Unix (`ifndef noerrnoiconv`) without those preproc guards, using `errno_location^` instead of `cerrno`, and using the respective error codes I found in my local `errno.h` from the MXE build of iconv